File: R/dictionary.R
#' Look up EAVS variables and concepts
#'
#' Searches the tidyeavs dictionary, which maps each survey year's raw variable
#' codes to stable concept names. Use it to answer "what is `C9a`?", to find
#' the code for a concept in a given year, or to see which years collected an
#' item.
#'
#' @param query A string matched (case-insensitively) against concept names,
#'   labels, raw codes, and EPI names. `NULL` (default) returns everything,
#'   subject to `year` and `section`.
#' @param year Restrict to one or more survey years.
#' @param section Restrict to one or more survey sections (`"A"` through `"F"`,
#'   or `"id"` for identifier columns).
#' @param dictionary The dictionary to search. Defaults to the bundled
#'   [eavs_dictionary]; pass your own to search an extended crosswalk.
#'
#' @return A tibble of matching dictionary rows (see [eavs_dictionary]).
#' @export
#'
#' @examples
#' # What does C9a measure, and is it C9a in every year?
#' eavs_items("C9a")
#'
#' # All mail-ballot concepts:
#' eavs_items(section = "C")
#'
#' # The code for mail ballots rejected, by year:
#' eavs_items("mail_rejected")
eavs_items <- function(query = NULL, year = NULL, section = NULL,
                       dictionary = NULL) {
  dict <- resolve_dictionary(dictionary)
  if (!is.null(year)) {
    dict <- dict[dict$year %in% year, , drop = FALSE]
  }
  if (!is.null(section)) {
    dict <- dict[tolower(dict$section) %in% tolower(section), , drop = FALSE]
  }
  if (!is.null(query)) {
    hay <- tolower(paste(
      dict$concept, dict$concept_label, dict$codebook_label,
      dict$code, dict$epi_name
    ))
    dict <- dict[grepl(tolower(query), hay, fixed = TRUE), , drop = FALSE]
  }
  tibble::as_tibble(dict)
}

# The dictionary to use: the supplied one, or the bundled dataset.
resolve_dictionary <- function(dictionary = NULL) {
  if (!is.null(dictionary)) {
    return(dictionary)
  }
  tryCatch(
    eavs_dictionary,
    error = function(e) {
      cli::cli_abort(c(
        "The bundled EAVS dictionary isn't available.",
        i = "Reinstall tidyeavs, or pass {.arg dictionary}."
      ))
    }
  )
}

# Dictionary rows for one year that have a raw code (i.e., were collected).
dict_for_year <- function(year, dictionary = NULL) {
  dict <- resolve_dictionary(dictionary)
  rows <- dict[dict$year == year & !is.na(dict$code), , drop = FALSE]
  if (nrow(rows) == 0) {
    cli::cli_abort("The dictionary has no entries for {year}.")
  }
  rows
}

# Resolve a single year from an explicit argument or a `year` column.
resolve_year <- function(data, year) {
  if (!is.null(year)) {
    if (length(year) != 1) {
      cli::cli_abort("{.arg year} must be a single year.")
    }
    return(year)
  }
  if ("year" %in% names(data)) {
    yy <- unique(data$year)
    if (length(yy) == 1) {
      return(yy)
    }
  }
  cli::cli_abort(c(
    "Could not tell which survey year {.arg data} is.",
    i = "Pass {.arg year}, or include a single-valued {.field year} column."
  ))
}

File: py/src/tidyeavs/_dictionary.py
"""Looking up EAVS variables and concepts."""

from __future__ import annotations

from typing import Iterable

import pandas as pd

from . import _metadata


def get_dictionary(dictionary: pd.DataFrame | None = None) -> pd.DataFrame:
    """The dictionary to use: the one supplied, or the shared crosswalk."""
    return _metadata.dictionary() if dictionary is None else dictionary


def items(
    query: str | None = None,
    year: int | Iterable[int] | None = None,
    section: str | Iterable[str] | None = None,
    dictionary: pd.DataFrame | None = None,
) -> pd.DataFrame:
    """Search the crosswalk that maps raw variable codes to concept names.

    Use it to answer "what is ``C9a``?", to find the code for a concept in a
    given year, or to see which years collected an item. ``query`` is matched
    case-insensitively against concept names, labels, raw codes, and EPI names.

    Because the mapping is explicit, the renumbering traps become visible rather
    than silent: ``items("uocava_rejected")`` shows the code is ``B24a`` in 2024
    but ``B18a`` in 2018 through 2022, and ``items("B18a")`` shows that ``B18a``
    in 2024 is a different item entirely.
    """
    frame = get_dictionary(dictionary)

    if year is not None:
        wanted = [year] if isinstance(year, int) else list(year)
        frame = frame[frame["year"].isin(wanted)]

    if section is not None:
        wanted_sections = [section] if isinstance(section, str) else list(section)
        lowered = [s.lower() for s in wanted_sections]
        frame = frame[frame["section"].str.lower().isin(lowered)]

    if query is not None:
        haystack = (
            frame[["concept", "concept_label", "codebook_label", "code", "epi_name"]]
            .fillna("")
            .agg(" ".join, axis=1)
            .str.lower()
        )
        frame = frame[haystack.str.contains(query.lower(), regex=False)]

    return frame.reset_index(drop=True)


def dict_for_year(
    year: int, dictionary: pd.DataFrame | None = None
) -> pd.DataFrame:
    """Dictionary rows for one year that have a code, i.e. were collected."""
    frame = get_dictionary(dictionary)
    rows = frame[(frame["year"] == year) & frame["code"].notna()]
    if rows.empty:
        raise ValueError(f"The dictionary has no entries for {year}.")
    return rows


def resolve_year(data: pd.DataFrame, year: int | None) -> int:
    """Work out which single survey year ``data`` is."""
    if year is not None:
        return int(year)
    if "year" in data.columns:
        unique = pd.unique(data["year"].dropna())
        if len(unique) == 1:
            return int(unique[0])
    raise ValueError(
        "Could not tell which survey year this data is. Pass year=, or include "
        "a single-valued 'year' column."
    )
